### ISSUE TITLE
MirroredStrategy doesn't use GPUs

### ISSUE THREAD

**Issue by czabo** [tensorflow/tensorflow#26020]
**System information**

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- TensorFlow installed from (source or binary): pip install tensorflow-gpu==1.12
- TensorFlow version (use command below): 1.12
- Python version: 3.5.2
- CUDA/cuDNN version: CUDA 9 / cuDNN 7.3
- GPU model and memory: 2 x Nvidia Titan X

**Describe the current behavior**

I was working on rewriting a script from the queue/threading approach to tf.data.Dataset approach of providing data. I got really nice throughput of data with over >90% util of GPUs.  Now that I have rewritten it, when starting the training with MirroredStrategy the GPUs are not used at all and I get the following output:

```
INFO:tensorflow:Device is available but not used by distribute strategy: /device:CPU:0
INFO:tensorflow:Device is available but not used by distribute strategy: /device:XLA_GPU:0
INFO:tensorflow:Device is available but not used by distribute strategy: /device:XLA_CPU:0
INFO:tensorflow:Configured nccl all-reduce.
INFO:tensorflow:Calling model_fn.
...
```

At this point I am thinking there is some issue with TF1.12.

**Code to reproduce the issue**

Here is basically the structure of my code. I tried out different things like train directly with tf.keras.fit with multi_gpu_model() but it didn't work out. Basically I am trying to reproduce the functionality of my RandomShuffleQueue() I had before with multiple threads filling up the queue. 

```
model = models.Model(inputs=input, outputs=y_output)
optimizer = tf.train.AdamOptimizer(LEARNING_RATE)
# model = utils.multi_gpu_model(model, gpus=NUM_GPUS, cpu_relocation=True)
model.compile(loss=lossFunc, optimizer=optimizer)

def generator(n):
    while True:
        try:
            imgBatch = []
            
            ...
            
            yield imgBatch
                
        except ValueError:
            pass
        
def get_generator(n):
    return partial(generator, n)

def dataset(n):
    return tf.data.Dataset.from_generator(get_generator(n), output_types=[(tf.float32, tf.float32)], output_shapes=[(tf.TensorShape([None,None,1]),tf.TensorShape([None,None,1]))

def input_fn():
    ds = tf.data.Dataset.range(len(dataSets)).apply(tf.data.experimental.parallel_interleave(dataset, cycle_length=len(dataSets), sloppy=True))
    ds = ds.map(map_func=lambda imgBatch: processImage(img,lbl), num_parallel_calls=12)
    ds = ds.shuffle(SHUFFLE_SIZE)
    ds = ds.batch(BATCH_SIZE)
    ds = ds.prefetch(1)
    return ds
                                                                                                                    
strategy = tf.contrib.distribute.MirroredStrategy(num_gpus=NUM_GPUS)
config = tf.estimator.RunConfig(train_distribute=strategy)
estimator = tf.keras.estimator.model_to_estimator(model,
                                                  config=config)
                                                                                                                    
estimator.train(lambda:input_fn())
```

Any help would be greatly appreciated since I'm stuck on it since a while now.

**Comment by guptapriya** [comment]
That output looks fine and is expected. The devices it should be using are "/device:GPU:0" etc which it doesn't complain about.
Why do you think GPUs are not being used?

**Comment by czabo** [comment]
I see. The info seemed strange to me. But also I I don't see a lot of util on my GPUs - to be exact it is about 5%/0%, peaking at around 27%/10% sometimes in spikes here and there. In contrast when using the old queue/threaded approach (using a RandomShuffleQueue) instead of tf.data.Dataset, I had a much higher util on the GPUs - about 50% on each GPU without any break. 

So I'm thinking there must be something wrong with either my input pipeline or the initialization of the GPUs. Any ideas? Maybe there is some issue on how I use the Dataset API and there is some bottleneck regarding creating the data. I am creating multiple Generators to simulate multi-threaded providing of data for shuffling. Appreciate any help since I'm struggling with that for quite some time now. 

Also any ideas how i could debug that issue?

**Comment by czabo** [comment]
Is there actually a proper way to provide the Dataset Shuffle with data in a multi-threaded way? The solution that I did feels kind of hacky and not proper. Might also be the reason it doesn't work as intended.

**Comment by byronyi** [comment]
I believe it's an input pipeline issue. Could you check out the [performance guide](https://www.tensorflow.org/guide/performance/datasets)?

**Comment by isaprykin** [comment]
You could monitor `$ nvidia-smi` during your training to see whether GPUs are used as well.

Please re-open if you think it's not an input pipeline issue and still suspect TF not utilizing available GPUs.
